torch/_custom_op/autograd.py

```diff
@@ -0,0 +1,222 @@
+import torch
+import torch.utils._pytree as pytree
+from collections import namedtuple
+
+
+# NOTE [CustomOp autograd kernel indirection]
+def autograd_kernel_indirection(custom_op):
+    # TODO(#101191): Use the actual C++ autograd not implemented fallback,
+    # or change the default autograd fallback to the autograd not implemented fallback.
+    def autograd_not_implemented(*args, **kwargs) -> None:
+        if pytree.tree_any(
+            lambda x: isinstance(x, torch.Tensor) and x.requires_grad, (args, kwargs)
+        ):
+            raise RuntimeError("Autograd has not been implemented for operator")
+        guard = torch._C._AutoDispatchBelowAutograd()
+        try:
+            return custom_op(*args, **kwargs)
+        finally:
+            del guard
+
+    def inner(*args, **kwargs):
+        if custom_op._has_impl('autograd'):
+            kernel = custom_op._get_impl('autograd').func
+            return kernel(*args, **kwargs)
+        if custom_op._has_impl('save_for_backward') or custom_op._has_impl('backward'):
+            missing = (
+                'save_for_backward' if custom_op._has_impl('backward')
+                else 'backward'
+            )
+            found = 'save_for_backward' if missing == 'backward' else 'backward'
+            loc = custom_op._get_impl(found).location
+            raise RuntimeError(
+                f"We found a '{found}' registration for {custom_op} at "
+                f"{loc} but were unable to find a '{missing}' registration. "
+                f"To use the CustomOp API to register a backward formula, "
+                f"please provide us both a backward function and a "
+                f"'save for backward' function via `impl_backward` and "
+                f"`impl_save_for_backward` respectively.")
+        return autograd_not_implemented(*args, **kwargs)
+    return inner
+
+
+def construct_autograd_kernel(
+        schema,
+        output_differentiability,
+        forward_op,
+        save_for_backward_fn,
+        backward_fn):
+
+    def apply(*args):
+        flat_args, spec = pytree.tree_flatten(args)
+
+        def forward(ctx, *flat_args):
+            ctx.set_materialize_grads(True)
+            args = pytree.tree_unflatten(flat_args, spec)
+            guard = torch._C._AutoDispatchBelowAutograd()
+            try:
+                output = forward_op(*args)
+            finally:
+                del guard
+
+            # We use the info about args to give better error messages in backward
+            args_info = namedtuple_args(
+                schema, pytree.tree_map(lambda arg: type(arg), args))
+
+            save_for_backward_fn_inputs = namedtuple_args(schema, args)
+            to_save = save_for_backward_fn(save_for_backward_fn_inputs, output)
+
+            save_pytree_for_backward(ctx, (to_save, args_info))
```

Comment:
nit we should be able to skip this when grad mode is disabled?

Reply:
Sure

Reply:
It is a little complicated to do this because autograd.Function's forward sets grad_mode to False. This means that we would need to shepherd the knowledge of if grad mode is disabled or not from somewhere else.

Since this is a nit I'm just going to add this to the wishlist.